Shared/Samples/Trace utility network/TraceUtilityNetworkView.Model.swift

```diff
@@ -0,0 +1,357 @@
+                    }
+                case .edge:
+                    if let line = GeometryEngine.makeGeometry(from: geometry, z: nil) as? Polyline {
+                        element.fractionAlongEdge = GeometryEngine.polyline(
+                            line,
+                            fractionalLengthClosestTo: mapPoint,
+                            tolerance: -1
+                        )
+                        Task {
+                            await updateUserHint(
+                                withMessage: String(format: "fractionAlongEdge: %.3f", element.fractionAlongEdge)
+                            )
+                        }
+                        add(element, at: mapPoint)
+                    }
+                @unknown default:
+                    return
+                }
+            }
+        }
+        
+        /// Identifies the first discoverable feature at the provided screen point.
+        /// - Parameters:
+        ///   - screenPoint: The location on the screen where the identify operation is desired.
+        ///   - proxy: The map view proxy to perform the identify operation with.
+        /// - Returns: The first discoverable feature or `nil` if none were identified.
+        func identifyFeatureAt(_ screenPoint: CGPoint, with proxy: MapViewProxy) async -> ArcGISFeature? {
+            try? await proxy.identifyLayers(
+                screenPoint: screenPoint,
+                tolerance: 10
+            ).first?.geoElements.first as? ArcGISFeature
+        }
+        
+        /// Sets the pending trace parameters with the provided trace type.
+        /// - Parameter type: The trace type.
+        func setTraceParameters(ofType type: UtilityTraceParameters.TraceType) {
+            pendingTraceParameters = UtilityTraceParameters(
+                traceType: type,
+                startingLocations: []
+            )
+            pendingTraceParameters?.traceConfiguration = mediumVoltageRadial?.defaultTraceConfiguration
+            tracingActivity = .settingPoints(pointType: .start)
+        }
+        
+        /// Resets all of the important stateful values for when a trace is cancelled or completed.
+        func reset() {
+            map.operationalLayers.forEach { ($0 as? FeatureLayer)?.clearSelection() }
+            points.removeAllGraphics()
+            pendingTraceParameters = nil
+            tracingActivity = .none
+        }
+        
+        /// Performs important tasks including adding credentials, loading and adding operational layers.
+        func setup() async {
+            do {
+                try await ArcGISEnvironment.authenticationManager.arcGISCredentialStore.add(.publicSample)
+                try await network?.load()
+            } catch {
+                await updateUserHint(withMessage: "An error occurred while loading the network.")
+                return
+            }
+            featureLayerURLs.forEach { url in
```

Comment:
Here is an idea you may take or ignore. Instead of adding the 2 feature layers to the map again, you can retrieve them from the portal item itself, such as 

```swift
private var featureLayers: [FeatureLayer] {
    map.operationalLayers
        .compactMap {
            ($0 as? GroupLayer)?.layers.compactMap { $0 as? FeatureLayer }
        }
        .flatMap { $0 }
}
```

Benefits:

- No need to create additional replicas of these 2 layers
- All the feature layers come in the order of how they are authored in the portal item
- Can get rid of a few variables such as `featureService` and `featureLayerURLs`

Cons:

- This group layer structure might not apply to users' own portal items
- Takes time to loop through layers. If an item contains many layers, it is slow

Reply:
Not sure I fully understand the suggestion here, can you clarify a bit? 

> Instead of adding the 2 feature layers to the map again

I think these need to be added no matter what?

Reply:
The portal item https://www.arcgis.com/home/item.html?id=471eb0bf37074b1fbb972b1da70fb310 contains a group layer, which bundles all the feature layers.

In theory, we can directly use the feature layers in that group layer because they are already part of the map. But that involves filtering them from the `operationalLayers` and create separate references. Which is why we added them as duplicates using the service feature table (with layer ID) URLs, just to display them and makes it easy to get to the 2 specific layers.

What I suggest is to retrieve the 2 feature layers from the group layer, so no duplicate layers that use the same feature table in the map. But there are some drawbacks as I listed, so if you think the current approach is good, I don't think it needs to be changed.

Reply:
I see. I think I prefer the way it's currently written as I find it a bit more digestible and [it's similar to the Objective C approach](https://github.com/Esri/arcgis-runtime-samples-ios/blob/b2694f9e5a97cd81ea042c1078708215ab4f3f48/arcgis-ios-sdk-samples/Utility%20network/Trace%20utility%20network/TraceUtilityNetworkViewController.swift#L72-L102).